Replaced PHP8-only str_contains method.

reense opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
This PR removes the PHP8-only method str_contains and replaces it with the traditional stripos() check with false, which is available since PHP5.

The reason for using this older method are the Magento 2.4 system requirements: Magento 2.4 supports PHP7.4 which does not contain the str_contains method [source](https://devdocs.magento.com/guides/v2.4/install-gde/system-requirements.html). The latest version of Magento, 2.4.3, does not even support PHP 8 [source](https://devdocs.magento.com/guides/v2.4/release-notes/open-source-2-4-3.html).

This issue has to be fixed, since it breaks upgrading a Magento instance on PHP 7.4. (see Manual error-reproducing scenario)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#33680
2. Fixes https://github.com/magento/magento2/issues/33755
3. Fixes https://github.com/magento/magento2/issues/33945
### Manual error-reproducing scenario
1. Create a Magento installation on any version lower than 2.4.3.
2. Use PHP 7.4.
3. Run `composer require magento/product-community-edition=2.4.3 --no-update`
4. Run `composer update`
5. The update fails; saying `call to undefined method str_contains().`

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

Commit:
Replaced PHP8-only str_contains method.

magento-engcom-team commented:
Hi @bgorski, thank you for the review.
[ENGCOM-9192](https://jira.corp.magento.com/browse/ENGCOM-9192) has been created to process this Pull Request

magento-engcom-team commented:
@reense thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

bgorski commented:
@magento run all tests

ihor-sviziev commented on lib/internal/Magento/Framework/Filesystem/Directory/DenyListPathValidator.php:
```diff
@@ -71,7 +71,7 @@ public function validate(
         foreach ($this->fileDenyList as $file) {
             $baseName = pathinfo($actualPath, PATHINFO_BASENAME);
-            if (str_contains($baseName, $file) || preg_match('#' . "\." . $file . '#', $fullPath)) {
+            if (stripos($baseName, $file) !== false || preg_match('#' . "\." . $file . '#', $fullPath)) {
```
As we already have the `symfony/polyfill-php80` in composer.lock file, I think it's better to add the `symfony/polyfill-php80` as a requirement to composer.json and use a new function.
https://github.com/magento/magento2/blob/9ae286d8a747f40fa0c89944f4aafc5cf4a5e9bb/composer.lock#L6645-L6727

If you still believe it should replaced with old function - please replace it with the equivalent:
`'' === $needle || false !== strpos($haystack, $needle);`
https://github.com/symfony/polyfill-php80/blob/1100343ed1a92e3a38f9ae122fc0eb21602547be/Php80.php#L93

PS: Currently, there is adding PHP 8 support in progress

reense replied:
@ihor-sviziev  I have updated this pull request to reflect your suggestions.
I get that you guys are making progress on creating PHP 8.0 support! however, for now, PHP7.4 is still supported so I think using a function that is not in every PHP release is not a very good idea (for the time being).

Commit:
Check if needle is empty before using strpos.

ihor-sviziev commented on lib/internal/Magento/Framework/Filesystem/Directory/DenyListPathValidator.php:
```diff
@@ -71,7 +71,7 @@ public function validate(
         foreach ($this->fileDenyList as $file) {
             $baseName = pathinfo($actualPath, PATHINFO_BASENAME);
-            if (str_contains($baseName, $file) || preg_match('#' . "\." . $file . '#', $fullPath)) {
+            if ('' === $needle || false !== strpos($haystack, $needle) || preg_match('#' . "\." . $file . '#', $fullPath)) {
```
```suggestion
            if ('' === $file || false !== strpos($baseName, $file) || preg_match('#' . "\." . $file . '#', $fullPath)) {
```

ihor-sviziev committed:
Replaced PHP8-only str_contains method

Fix after code review

reense commented on lib/internal/Magento/Framework/Filesystem/Directory/DenyListPathValidator.php:
```diff
@@ -71,7 +71,7 @@ public function validate(
 
         foreach ($this->fileDenyList as $file) {
             $baseName = pathinfo($actualPath, PATHINFO_BASENAME);
-            if (str_contains($baseName, $file) || preg_match('#' . "\." . $file . '#', $fullPath)) {
```
@ihor-sviziev welp.. sorry I was (apparently) not paying enough attention when performing this change.

engcom-Hotel commented:
✖️ QA Failed

Hello @reense,

Thanks for the contribution!

I have tried to reproduce the issue but was unable to do so. I have tried the below steps in order to reproduce the issue:
1. Install Magento 2.4.2-P1 
2. Run command `composer require magento/product-community-edition=2.4.3 --no-update`
3. And then `composer update`
4. I have PHP 7.4 version installed in my system. Please refer the below screenshot for reference:
<img width="620" alt="Screenshot 2021-08-31 at 6 16 52 PM" src="https://user-images.githubusercontent.com/51681618/131505098-caef51da-efcf-44e4-a4cf-518881034446.png">

After following the above steps, the `composer update` command is running successfully. It might be because `symfony/polyfill-php80` library is already installed. It has the `str_contains` method in the bootstrap.php file.

Please let us know any further steps in order to reproduce the issue.

Thanks

engcom-Hotel reviewed (changes requested):
Unable to reproduce the issue. Please follow my comments on this [link](https://github.com/magento/magento2/pull/33886#issuecomment-909207080).

nbennett25 commented:
@engcom-Hotel I am able to reproduce the error with 2.4.2-p2, but was able to apply patch AC-384__Fix_Incompatible_PHP_Method__2.4.3_ce with success.

ihor-sviziev commented:
@engcom-Hotel, There is the official hotfix for this issue
https://support.magento.com/hc/en-us/articles/4408021533069-Adobe-Commerce-upgrade-2-4-3-2-3-7-p1-PHP-Fatal-error-Hotfix

so you can check internal docs maybe of how to reproduce it.

ihor-sviziev committed:
Replaced PHP8-only str_contains method

Update patch to be equal to the hotfix in
https://support.magento.com/hc/en-us/articles/4408021533069-Adobe-Commerce-upgrade-2-4-3-2-3-7-p1-PHP-Fatal-error-Hotfix

ihor-sviziev commented:
I just updated the PR to be equal to the official hotfix described here:
https://support.magento.com/hc/en-us/articles/4408021533069-Adobe-Commerce-upgrade-2-4-3-2-3-7-p1-PHP-Fatal-error-Hotfix
@magento run all tests

magento-engcom-team commented:
Hi @Den4ik, thank you for the review.
[ENGCOM-9192](https://jira.corp.magento.com/browse/ENGCOM-9192) has been created to process this Pull Request

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9192](https://jira.corp.magento.com/browse/ENGCOM-9192) has been created to process this Pull Request

engcom-Alfa commented:
I tried to reproduce the issue using 2.4-develop and could not reproduce at all!
There is a way I found out to check the _str_contains()_ function compatibility
1. Disabled the str_contains() if statement block completely from the _bootstrap.php_ class, then executed the upgrade command.
 2. Definitely we get the error as defined in this PR report.
 Well, updated PR code changes will help in resolving the issue as well where we have to reset the  _bootstrap.php_ class changes before the re execution!
@ihor-sviziev @reense @engcom-Hotel

engcom-Alfa commented:
@magento give me test instance
